### PR TITLE
add nanolyse module (from nanoseq modules)

### PR DESCRIPTION
**Pull request by yuukiiwa**

### CODE REVIEW

File: software/nanolyse/main.nf [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+process NANOLYSE {
+    tag "$meta.id"
+    label 'process_low'
+    publishDir "${params.outdir}",
+        mode: params.publish_dir_mode,
+        saveAs: { filename -> saveFiles(filename:filename, options:params.options, publish_dir:getSoftwareName(task.process), meta:meta, publish_by_meta:['id']) }
+
+    conda     (params.enable_conda ? "bioconda::nanolyse=1.2.0" : null)
+    container "quay.io/biocontainers/nanolyse:1.2.0--py_0"
```
Comment (KevinMenden) [inline software/nanolyse/main.nf:15]: This is missing the singularity container, it should be something like this:

```
    conda (params.enable_conda ? "bioconda::bcftools=1.11" : null)
    if (workflow.containerEngine == 'singularity' && !params.singularity_pull_docker_container) {
        container "https://depot.galaxyproject.org/singularity/bcftools:1.11--h7c999a4_0"
    } else {
        container "quay.io/biocontainers/bcftools:1.11--h7c999a4_0"
    }
```

File: software/nanolyse/main.nf [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+process GET_NANOLYSE_FASTA {
+    output:
+    path "*fasta.gz"  , emit: nanolyse_fasta
+
+    script:
+    """
+    wget https://github.com/wdecoster/nanolyse/raw/master/reference/lambda.fasta.gz
+    """
+}
```
Comment (KevinMenden) [inline software/nanolyse/main.nf:42]: What is this used for? This should probably not be in the module.

File: software/nanolyse/meta.yml [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+            Groovy Map containing sample information
+            e.g. [ id:'test', single_end:false ]
+    - fastq:
+        type: file
+        description: |
+            Basecalled reads in FASTQ.GZ format
+    - nanolyse_fasta:
+        type: file
+        description: |
+            A reference fasta file against which to filter.
```
Comment (KevinMenden) [inline software/nanolyse/meta.yml:24]: would be nice to have the pattern of the files here :)

**Review by KevinMenden — CHANGES_REQUESTED** [review]
Thank you!
Only some small things need to be addressed :+1:

File: software/nanolyse/functions.nf [hunk 4]
```diff
@@ -0,0 +1,70 @@
+/*
+ * -----------------------------------------------------
```
[the remark this answers is not in the document]
Comment (yuukiiwa) [reply software/nanolyse/functions.nf:2]: `functions.nf` is now using the new template

File: software/nanolyse/main.nf (hunk 1 above)
Comment (yuukiiwa) [reply software/nanolyse/main.nf]: I have added the singularity container for nanolyse

File: software/nanolyse/meta.yml (hunk 3 above)
Comment (yuukiiwa) [reply software/nanolyse/meta.yml]: I have added the patterns

File: software/nanolyse/main.nf (hunk 2 above)
Comment (yuukiiwa) [reply software/nanolyse/main.nf:46]: Hi @KevinMenden, this process is for staging the `nanolyse` fasta file provided by `nanolyse`. This fasta file is only used by the `NANOLYSE` process, so I'm not sure whether I should separate this `GET_NANOLYSE_FASTA` process into another module.

Comment (KevinMenden) [reply software/nanolyse/main.nf:46]: Hi @yuukiiwa I would say you should separate it. 

To me this look rather like something that would be a local module :thinking: not sure whether we need to include it in nf-core/modules.
Although you could make it a submodule of nanlyse.

Comment (yuukiiwa) [reply software/nanolyse/main.nf:46]: Thanks! I think I will remove it here and add it locally for the test.

Comment (yuukiiwa) [reply software/nanolyse/main.nf]: Hi @KevinMenden, I have moved the `GET_NANOLYSE_FASTA` process to the `NANOLYSE` test as a local module. Thanks!

**Comment by KevinMenden** [comment]
Hi @yuukiiwa , thanks for fixing that! I think it's fine like this, but pinging @drpatelh to get a second opinion.

File: software/nanolyse/functions.nf [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+/*
+ * Function to save/publish module results
+ */
+def saveFiles(Map args) {
+    if (!args.filename.endsWith('.version.txt')) {
+        def ioptions = initOptions(args.options)
+        def path_list = [ ioptions.publish_dir ?: args.publish_dir ]
+        if (ioptions.publish_by_id) {
```
Comment (drpatelh) [inline software/nanolyse/functions.nf:45]: This is still the old `functions.nf`

File: software/nanolyse/main.nf [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+def options    = initOptions(params.options)
+
+process NANOLYSE {
+    tag "$meta.id"
+    label 'process_low'
+    publishDir "${params.outdir}",
+        mode: params.publish_dir_mode,
+        saveAs: { filename -> saveFiles(filename:filename, options:params.options, publish_dir:getSoftwareName(task.process), meta:meta, publish_by_meta:['id']) }
+
+    conda     (params.enable_conda ? "bioconda::nanolyse=1.2.0" : null)
```
Comment (drpatelh) [inline software/nanolyse/main.nf:14]:
```suggestion
    conda (params.enable_conda ? "bioconda::nanolyse=1.2.0" : null)
```

File: software/nanolyse/main.nf [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    input:
+    tuple val(meta), path(fastq)
+    path nanolyse_fasta
+
+    output:
+    tuple val(meta), path("*.fastq.gz")   ,emit: nanolyse_fastq
+    path "*.log"                          ,emit: nanolyse_log
+    path "*.version.txt"                  ,emit: version
```
Comment (drpatelh) [inline software/nanolyse/main.nf:28]:
```suggestion
    tuple val(meta), path("*.fastq.gz"), emit: fastq
    path "*.log"                       , emit: log
    path "*.version.txt"               , emit: version
```

File: software/nanolyse/main.nf [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    conda     (params.enable_conda ? "bioconda::nanolyse=1.2.0" : null)
+    if (workflow.containerEngine == 'singularity' && !params.singularity_pull_docker_container) {
+        container "https://depot.galaxyproject.org/singularity/nanolyse:1.2.0--py_0"
+    } else {
+        container "quay.io/biocontainers/nanolyse:1.2.0--py_0"
+    }
+
+    input:
+    tuple val(meta), path(fastq)
+    path nanolyse_fasta
```
Comment (drpatelh) [inline software/nanolyse/main.nf:23]:
```suggestion
    path fasta
```

File: software/nanolyse/main.nf [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    tuple val(meta), path("*.fastq.gz")   ,emit: nanolyse_fastq
+    path "*.log"                          ,emit: nanolyse_log
+    path "*.version.txt"                  ,emit: version
+
+    script:
+    """
+    gunzip -c $fastq | NanoLyse -r $nanolyse_fasta | gzip > ${meta.id}.clean.fastq.gz
+    cp NanoLyse.log ${meta.id}.nanolyse.log
+    NanoLyse --version &> nanolyse.version.txt
+    """
```
Comment (drpatelh) [inline software/nanolyse/main.nf:35]:
```suggestion
    script:
    def software = getSoftwareName(task.process)
    def prefix   = options.suffix ? "${meta.id}.${options.suffix}" : "${meta.id}"
    """
    gunzip -c $fastq | NanoLyse -r $fasta | gzip > ${prefix}.fastq.gz
    mv NanoLyse.log ${prefix}.nanolyse.log

    echo \$(NanoLyse --version 2>&1) | sed -e "s/NanoLyse //g" > ${software}.version.txt
    """
```

File: software/nanolyse/meta.yml [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+        type: map
+        description: |
+            Groovy Map containing sample information
+            e.g. [ id:'test', single_end:false ]
+    - fastq:
+        type: file
+        description: |
+            Basecalled reads in FASTQ.GZ format
+        pattern: "*.fastq.gz"
+    - nanolyse_fasta:
```
Comment (drpatelh) [inline software/nanolyse/meta.yml:22]:
```suggestion
    - fasta:
```

File: software/nanolyse/meta.yml [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+        description: |
+            A reference fasta file against which to filter.
+        pattern: "*.fasta"
+output:
+    - meta:
+        type: map
+        description: |
+            Groovy Map containing sample information
+            e.g. [ id:'test', single_end:false ]
+    - fastq:
```
Comment (drpatelh) [inline software/nanolyse/meta.yml:33]: Add entry for `log` here too.

File: software/nanolyse/meta.yml [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+        type: file
+        description: Reads with contaminants removed in FASTQ format
+        pattern: "*.clean.fastq.gz"
```
Comment (drpatelh) [inline software/nanolyse/meta.yml:36]:
```suggestion
        pattern: "*.fastq.gz"
```

File: tests/software/nanolyse/main.nf [hunk 13]
```diff
@@ -0,0 +1,24 @@
+#!/usr/bin/env nextflow
+
+nextflow.enable.dsl = 2
+
+include { NANOLYSE } from '../../../software/nanolyse/main.nf'    addParams( options: [:] )
+
+process GET_NANOLYSE_FASTA {
+    output:
+    path "*fasta.gz"  , emit: nanolyse_fasta
```
Comment (drpatelh) [inline tests/software/nanolyse/main.nf:9]:
```suggestion
    path "*fasta.gz", emit: fasta
```

File: tests/software/nanolyse/main.nf [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+    """
+}
+
+workflow test_nanolyse {
+    input = [ [ id:'test' ], // meta map
+              [ file(params.test_data['sarscov2']['nanopore']['test_fastq_gz'], checkIfExists: true)]
+            ]
+
+    GET_NANOLYSE_FASTA()
+    NANOLYSE ( input, GET_NANOLYSE_FASTA.out.nanolyse_fasta )
```
Comment (drpatelh) [inline tests/software/nanolyse/main.nf:23]:
```suggestion
    NANOLYSE ( input, GET_NANOLYSE_FASTA.out.fasta )
```

File: tests/software/nanolyse/test.yml [hunk 15]
```diff
@@ -0,0 +1,6 @@
+- name: nanolyse
+  command: nextflow run ./tests/software/nanolyse -entry test_nanolyse -c tests/config/nextflow.config
+  tags:
+    - nanolyse
+  files:
+    - path: ./output/nanolyse/test.clean.fastq.gz
```
Comment (drpatelh) [inline tests/software/nanolyse/test.yml:6]:
```suggestion
    - path: ./output/nanolyse/test.fastq.gz
```

**Review by drpatelh — DISMISSED** [review]
Looks better @yuukiiwa ! Still a few changes left :)

**Review by drpatelh — CHANGES_REQUESTED** [review]
Just clicking the button to request changes again :)

**Comment by yuukiiwa** [comment]
Thanks, @drpatelh! I will update you when I am done with the changes.

File: software/nanolyse/functions.nf (hunk 5 above)
Comment (yuukiiwa) [reply software/nanolyse/functions.nf]: I have updated this using the one from `nf-core/tools-dev`

**Comment by yuukiiwa** [comment]
Hi @drpatelh, `gunzip` raises an error because `{prefix}.fastq.gz` and `$fastq` were having the same name:
Screenshot 2021-05-05 at 1 45 59 PM
So I tried reverting `{prefix}.fastq.gz` back to `{prefix}.clean.fastq.gz` and changing `{prefix}.clean.fastq.gz` to `{prefix}.fastq.gz`, but it raises an error because the input and output files are having the same name:
Screenshot 2021-05-05 at 1 46 33 PM
Other than this, I have modified the PR according to your suggestions.
Thanks!

File: software/nanolyse/meta.yml (hunk 11 above)
Comment (yuukiiwa) [reply software/nanolyse/meta.yml:33]: I added an entry for `log`

File: software/nanolyse/main.nf [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+    output:
+    tuple val(meta), path("*.fastq.gz"), emit: fastq
+    path "*.log"                       , emit: log
+    path "*.version.txt"               , emit: version
+
+    script:
+    def software = getSoftwareName(task.process)
+    def prefix   = options.suffix ? "${meta.id}.${options.suffix}" : "${meta.id}"
+    """
+    gunzip -c $fastq | NanoLyse -r $fasta | gzip > ${prefix}.clean.fastq.gz
```
Comment (drpatelh) [inline software/nanolyse/main.nf:34]:
```suggestion
    gunzip -c $fastq | NanoLyse -r $fasta | gzip > ${prefix}.fastq.gz
```

File: tests/software/nanolyse/main.nf [hunk 17]
```diff
@@ -0,0 +1,24 @@
+#!/usr/bin/env nextflow
+
+nextflow.enable.dsl = 2
+
+include { NANOLYSE } from '../../../software/nanolyse/main.nf'    addParams( options: [:] )
```
Comment (drpatelh) [inline tests/software/nanolyse/main.nf:5]:
```suggestion
include { NANOLYSE } from '../../../software/nanolyse/main.nf' addParams( options: [suffix: 'clean'] )
```

**Comment by drpatelh** [comment]
Try those 2 suggestions @yuukiiwa . So the filename in the main module script should be left as vanilla as possible because you can customise the filename by passing the `suffix` via `options`.

**Comment by yuukiiwa** [comment]
Thank you so much, @drpatelh! I have committed your suggestions!

**Review by drpatelh — APPROVED** [review]
LGTM! Thanks @yuukiiwa

**Pull request merged**